### PR TITLE
tqdm_logger update (#296)

### LINKED ISSUE #296: Discussion about ProgressBar and metrics display
Hello, I would like to discuss the feature in `contrib.handlers.ProgressBar` that displays metrics on each `Events.ITERATION_COMPLETED` as a postfix of the bar. For some context, please read https://github.com/pytorch/ignite/pull/295.

I think the `ProgressBar` and this feature are essential to the package, as it provides a concise way to have a nice training and evaluation experience (replacing the old way of spamming numbers in the console), and is appealing to beginners.

Here is some of the issues I found.

### Problem
On the doc, the given example is
```python
pbar = ProgressBar()
pbar.attach(trainer, ['loss'])
```

The `metrics` parameter (here `['loss']`) works with `Metric`s that have been attached to the engine, **AND that rewired `self.computed` (which updates the `state.metrics` dict of the engine) on `Events.ITERATION_COMPLETED` instead of `Events.EPOCH_COMPLETED`**. The second part is unintuitive, and not documented. The current document suggests that this might work with all metrics, but this raises a `KeyError` when the dict has not been updated before the end of the first iteration.
Currently, this freature only works with `RunningAverage`.

The problem has been raised by @vfdev-5 in https://github.com/pytorch/ignite/pull/256, but I did not see the discussion go anywhere, I would like for it to continue.

### Proposed solutions

1. The documentation should be updated to reflect the fact that this feature will not work on all of the built-in metrics but `RunningAverage`.

2. The error message for the `KeyError` is `metrics not found in engine.state.metrics` and could be modified to hint for what the issue might be.

3. a Maybe the `Metric` class hierarchy should be updated, to distinguish metrics that update the engine state every epoch, and those that update it every iteration. The `ProgressBar` could then only accept the latter one. 

3. b Or maybe all of the current built-in metrics could update the state of the engine at each iteration (on `self.update` instead of on `self.completed`. This will add computation at each epoch. The progress bar would then work on all metrics.

### Problem
Have a look at the doc example again :
```python
pbar = ProgressBar()
pbar.attach(trainer, ['loss'])
```
This suggests that we can print the training loss, that updates at each iteration.

However, we can't. When we are using a trainer created with `engine.create_supervised_trainer`, the `process_function` returns the loss for the processed batch. I think this is the most intuitive way to do it even for custom training engines.
The `Metric`s that can be attached are computed from the output of this function (`y_pred, y` for an evaluator for example), which may pass through a transform function. You see where I am going: the output of the trainer is already the loss itself, we do not need any further computation.

If we want to print the training loss, we have to attach a `Metric` that does nothing but moves the already computed loss into the `state.metrics` dict of the engine at every iteration. See https://github.com/pytorch/ignite/pull/295 for the `Identity` metric. This is not intuitive, we should be able to display the training loss out of the box.

### Proposed solutions

1. Built-in Identity metric, see https://github.com/pytorch/ignite/pull/295 for the code.

2. Modify `ProgressBar` to allow the display in the postfix the value of `engine.state.output`
    The `attach` method could then be 
    ```python
    attach(self, engine, metric_names=None, display_output=False)
    ```
    with the output looking like 
    `Epoch 1: [59/689]   9%|▊                 , output=3.54e-02 [00:04<00:50]`
3. Modify `ProgressBar` to allow the display in the postfix the value of any custom variables in the `engine.state` object, including the output.

---
Thanks for reading, sorry for the long issue and the english mistakes.

### PR DESCRIPTION
**Pull request by miguelvr**
Fixes #296 

Description:
added `persist` option and `output_transform` argument

Check list:
* [x] New tests are added
* [x] New doc strings: text and/or example code are in RST format
* [x] Documentation is updated

### CODE REVIEW

**Comment by miguelvr** [comment]
@vfdev-5 please have a look :)

File: ignite/contrib/handlers/tqdm_logger.py [hunk 1]
```diff
@@ -10,47 +10,84 @@ class ProgressBar:
...
+        Simple progress bar
+
+        .. code-block:: python
+
+            trainer = create_supervised_trainer(model, optimizer, loss)
+
+            pbar = ProgressBar()
+            pbar.attach(trainer)
+
+        Attach metrics that already has been computed (such as `RunningAverage`)
```
Comment (milesial) [inline ignite/contrib/handlers/tqdm_logger.py]:
```suggestion
        Attach metrics that already have been computed at `ITERATION_COMPLETED` (such as `RunningAverage`)
```

This better describes it IMO, as the metric isn't already computed at attach time, but should have updated the `state.metrics` of the engine before the bar updates, at ITERATION_COMPLETED

Comment (miguelvr) [reply ignite/contrib/handlers/tqdm_logger.py]: just fixed that :) thanks!

File: ignite/contrib/handlers/tqdm_logger.py [hunk 2]
```diff
@@ -65,16 +102,22 @@ def log_message(message):
...
 
-    def attach(self, engine, metric_names=None):
+    def attach(self, engine, metric_names=None, output_transform=None):
         """
         Attaches the progress bar to an engine object
 
         Args:
             engine (Engine): engine object
             metric_names (list): (Optional) list of the metrics names to log as the bar progresses
+            output_transform (function): (Optional) function to transform the engine output into a
```
Comment (vfdev-5) [inline ignite/contrib/handlers/tqdm_logger.py]:
```suggestion
            output_transform (Callable, optional) a function to transform the engine output into a
```

File: ignite/contrib/handlers/tqdm_logger.py [hunk 3]
```diff
@@ -10,47 +10,84 @@ class ProgressBar:
...
         if metric_names is not None:
             if not all(metric in engine.state.metrics for metric in metric_names):
                 self._close(engine)
                 raise KeyError("metrics not found in engine.state.metrics")
 
-            metrics = {name: '{:.2e}'.format(engine.state.metrics[name]) for name in metric_names}
+            metrics.update({name: '{:.2e}'.format(engine.state.metrics[name]) for name in metric_names})
+
+        if output_transform is not None:
+            output_dict = output_transform(engine.state.output)
```
Comment (vfdev-5) [inline ignite/contrib/handlers/tqdm_logger.py]: IMO, to force the output as dict is a bit too strict and too verbose if user would like to display a single scalar, e.g. `loss.item()`. However this can be more interesting to display multiple losses without smoothing. Maybe we can let two possibilities for the output: dict and "not dict" which is then wrapped into a dict as : 
```python
output_dict = {"output": output_dict}
```

Comment (miguelvr) [reply ignite/contrib/handlers/tqdm_logger.py]: it is harder to check for the proper output type, however. My bet would be to accept either a scalar or a `dict`

**Comment by miguelvr** [comment]
> @miguelvr what about custom variables ?

what do you mean by that?

**Comment by vfdev-5** [comment]
> what do you mean by that?

@miguelvr take a look [here](https://github.com/pytorch/ignite/issues/296#issuecomment-431697231) at display custom variables.

Actually, I think we will need to do a similar stuff for RunningAverage to support custom variables too...

File: ignite/contrib/handlers/tqdm_logger.py (hunk 3 above)
Comment (vfdev-5) [reply ignite/contrib/handlers/tqdm_logger.py]: Actually, you are not checking if the output is printable, I think this is OK. So you can just additionally compare if the output is not dict and wrap it to the dict.

Comment (miguelvr) [reply ignite/contrib/handlers/tqdm_logger.py]: > > what do you mean by that?
> 
> @miguelvr take a look [here](https://github.com/pytorch/ignite/issues/296#issuecomment-431697231) at display custom variables.
> 
> Actually, I think we will need to do a similar stuff for RunningAverage to support custom variables too...

I did look at that, but I don't understand the usage of it. Can you give me an example of it being necessary?

**Comment by miguelvr** [comment]
> > what do you mean by that?
> 
> @miguelvr take a look [here](https://github.com/pytorch/ignite/issues/296#issuecomment-431697231) at display custom variables.
> 
> Actually, I think we will need to do a similar stuff for RunningAverage to support custom variables too...

I looked at that but still don't understand the usage. Can you give me an example of such a thing being necessary?

**Comment by vfdev-5** [comment]
@miguelvr for example in a custom processing_function user store aux output in the state:
```python
def processing_function(engine, batch):
    # ...
    engine.state.aux_out = aux_out
    # ...
    return loss_value
```
and would like to display it too. With such option in progress bar we do not force the user to return this aux output with loss value and then dispatch it with the progress bar.
What do you think ?

**Comment by miguelvr** [comment]
In that example you would solve the problem simply by returning the value itself... To me, it seems a bit too cumbersome to support that, since it is not preventing the user to accomplish the same result.

**Comment by vfdev-5** [comment]
I see your point. Okay, let's leave it without custom variable support.

But still I'm not convinced on the output_transform to return directly a `dict`. I think it is better to have two options as discussed earlier...

**Comment by miguelvr** [comment]
> I see your point. Okay, let's leave it without custom variable support.
> 
> But still I'm not convinced on the output_transform to return directly a `dict`. I think it is better to have two options as discussed earlier...

and what is the name displayed then?

**Comment by vfdev-5** [comment]
@miguelvr as I proposed above:
```python
if output_transform is not None:
     output_dict = output_transform(engine.state.output)
    if not isinstance(output_dict, dict): 
        output_dict = {"output": output_dict}
```

File: ignite/contrib/handlers/tqdm_logger.py [hunk 4]
```diff
@@ -65,16 +101,23 @@ def log_message(message):
...
         tqdm.write(message)
 
-    def attach(self, engine, metric_names=None):
+    def attach(self, engine, metric_names=None, output_transform=None):
         """
         Attaches the progress bar to an engine object
 
         Args:
             engine (Engine): engine object
             metric_names (list): (Optional) list of the metrics names to log as the bar progresses
```
Comment (vfdev-5) [inline ignite/contrib/handlers/tqdm_logger.py]: @miguelvr could you please change this too
```
metric_names (list): (Optional)
```
to 
```
metric_names (list, optional): 
```
thanks

**Comment by vfdev-5** [comment]
@jasonkriss could you please review and merge if OK

File: ignite/contrib/handlers/tqdm_logger.py [hunk 5]
```diff
@@ -10,47 +10,83 @@ class ProgressBar:
...
-            metrics = {name: '{:.2e}'.format(engine.state.metrics[name]) for name in metric_names}
+            metrics.update({name: '{:.2e}'.format(engine.state.metrics[name]) for name in metric_names})
+
+        if output_transform is not None:
+            output_dict = output_transform(engine.state.output)
+
+            if not isinstance(output_dict, dict):
+                output_dict = {"output": output_dict}
+
+            metrics.update({name: '{:.2e}'.format(value) for name, value in output_dict.items()})
```
Comment (TheCodez) [inline ignite/contrib/handlers/tqdm_logger.py]: Why is `{:.2e}` used here? Every _ignite_ examples uses the `{:.2f}` formatting, right?

Comment (miguelvr) [reply ignite/contrib/handlers/tqdm_logger.py]: Exponential notation generalizes better regardless of the type of experiment. If you keep .2f you risk not having enough resolution to check the output

Comment (TheCodez) [reply ignite/contrib/handlers/tqdm_logger.py]: Yes that true, that's why I use `{:.4f}` most of the time as I ike it more than the exponential notation. Maybe there should be a way to specify the format?

Comment (vfdev-5) [reply ignite/contrib/handlers/tqdm_logger.py]: @miguelvr I think we can change to `{:.4f}`. I would prefer.

**Comment by amitibo** [comment]
@miguelvr and @vfdev-5 what about adding the ability to direct the output of the progress bar to a logging stream: https://stackoverflow.com/a/41224909

**Comment by vfdev-5** [comment]
@amitibo sounds good, thanks for the idea! Please create a new issue for that. As we have some handlers (Engine, EarlyStopping, ...) that have their own loggers, we need to think about if their integration is necessary etc.

File: ignite/contrib/handlers/tqdm_logger.py (hunk 5 above)
Comment (miguelvr) [reply ignite/contrib/handlers/tqdm_logger.py]: @vfdev-5 I currently have a project where even that resolution wouldn't cover one of my variables. I still vouch for the exponential notation as it is what guarantees resolution over all the cases.

Comment (vfdev-5) [reply ignite/contrib/handlers/tqdm_logger.py]: @miguelvr ok, let's keep it `{:.2e}` to guarantee the resolution.

**Comment by jasonkriss** [comment]
LGTM. Thanks @miguelvr!
